Mathlib/Geometry/Manifold/Bordisms.lean

```diff
@@ -0,0 +1,202 @@
+
+instance {n : ℕ} {k : ℕ∞} {s : SingularNManifold X n k I} : CompactSpace s.M := s.compactSpace
+
+instance {n : ℕ} {k : ℕ∞} {s : SingularNManifold X n k I} : BoundarylessManifold I s.M :=
+  s.boundaryless
+
+/-- A map of topological spaces induces a corresponding map of singular n-manifolds. -/
+-- This is part of proving functoriality of the bordism groups.
+noncomputable def map (s : SingularNManifold X n k I)
+    {φ : X → Y} (hφ : Continuous φ) : SingularNManifold Y n k I where
+  f := φ ∘ s.f
+  hf := hφ.comp s.hf
+  dimension := s.dimension
+
+@[simp]
+lemma map_f (s : SingularNManifold X n k I) {φ : X → Y} (hφ : Continuous φ) :
+    (s.map hφ).f = φ ∘ s.f :=
+  rfl
+
+lemma map_comp (s : SingularNManifold X n k I)
+    {φ : X → Y} {ψ : Y → Z} (hφ : Continuous φ) (hψ : Continuous ψ) :
+    ((s.map hφ).map hψ).f = (ψ ∘ φ) ∘ s.f := by
+  simp [Function.comp_def]
+  rfl
+
+-- Let M' and W be real C^k manifolds.
+variable {E' E'' E''' H' H'' H''' : Type*}
+  [NormedAddCommGroup E'] [NormedSpace ℝ E'] [NormedAddCommGroup E'']  [NormedSpace ℝ E'']
+  [NormedAddCommGroup E'''] [NormedSpace ℝ E''']
+  [TopologicalSpace H'] [TopologicalSpace H''] [TopologicalSpace H''']
+
+variable {M' : Type*} [TopologicalSpace M'] [ChartedSpace H' M']
+  {I' : ModelWithCorners ℝ E' H'} [IsManifold I' k M']
+  [BoundarylessManifold I' M'] [CompactSpace M'] [FiniteDimensional ℝ E']
+
+variable (M I) in
+/-- If `M` is `n`-dimensional and closed, it is a singular `n`-manifold over itself.-/
+noncomputable def refl (hdim : finrank ℝ E = n) :
+    SingularNManifold M n k I where
+  dimension := hdim
+  f := id
+  hf := continuous_id
+
+/-- If `(N, f)` is a singular `n`-manifold on `X` and `M` another `n`-dimensional manifold,
+a smooth map `φ : M → N` induces a singular `n`-manifold structure `(M, f ∘ φ)` on `X`. -/
+noncomputable def comap [h : Fact (finrank ℝ E' = n)]
+    (s : SingularNManifold X n k I)
+    {φ : M' → s.M} (hφ : ContMDiff I' I n φ) : SingularNManifold X n k I' where
+  f := s.f ∘ φ
+  hf := s.hf.comp hφ.continuous
+  dimension := h.out
+
+@[simp]
+lemma comap_f [Fact (finrank ℝ E' = n)]
+    (s : SingularNManifold X n k I) {φ : M' → s.M} (hφ : ContMDiff I' I n φ) :
+    (s.comap hφ).f = s.f ∘ φ :=
+  rfl
+
+variable (M I) in
+/-- The canonical singular `n`-manifold associated to the empty set (seen as an `n`-dimensional
+manifold, i.e. modelled on an `n`-dimensional space). -/
+def empty [h: Fact (finrank ℝ E = n)] (M : Type*) [TopologicalSpace M] [ChartedSpace H M]
+    {I : ModelWithCorners ℝ E H} [IsManifold I k M] [IsEmpty M] : SingularNManifold X n k I where
+  M := M
+  dimension := h.out
+  f := fun x ↦ (IsEmpty.false x).elim
+  hf := by
+    rw [continuous_iff_continuousAt]
+    exact fun x ↦ (IsEmpty.false x).elim
+
+variable (M I) in
+/-- An `n`-dimensional manifold induces a singular `n`-manifold on the one-point space. -/
+def trivial [h: Fact (finrank ℝ E = n)] : SingularNManifold PUnit n k I where
```

Comment:
It's not really trivial, as all the topology and smooth structure of `M` is in there, as well as the standard bordism theory. `toPUnit`, maybe?